test/Microsoft.ML.Core.Tests/UnitTests/TestHosts.cs

```diff
@@ -55,23 +55,62 @@ public void TestCancellation()
                     do
                     {
                         index = rand.Next(hosts.Count);
-                    } while (hosts.ElementAt(index).Item1.IsCancelled || hosts.ElementAt(index).Item2 < 3);
-                    hosts.ElementAt(index).Item1.StopExecution();
+                    } while ((hosts.ElementAt(index).Item1 as ICancelableEnvironment).IsCanceled || hosts.ElementAt(index).Item2 < 3);
+                    (hosts.ElementAt(index).Item1 as ICancelableHost).StopExecution();
                     rootHost = hosts.ElementAt(index).Item1;
                     queue.Enqueue(rootHost);
                 }
                 addThread.Join();
                 while (queue.Count > 0)
                 {
                     var currentHost = queue.Dequeue();
-                    Assert.True(currentHost.IsCancelled);
+                    Assert.True((currentHost as ICancelableEnvironment).IsCanceled);
 
                     if (children.ContainsKey(currentHost))
                         children[currentHost].ForEach(x => queue.Enqueue(x));
                 }
             }
         }
 
+        [Fact]
+        public void TestCancellationRoot()
+        {
+            IHostEnvironment env = new MLContext(seed: 42);
+            var mainHost = env.Register("Main");
+            var children = new ConcurrentDictionary<IHost, List<IHost>>();
+            var hosts = new BlockingCollection<Tuple<IHost, int>>();
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("1"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("2"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("3"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("4"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("5"), 1));
+            Random rand = new Random();
```

Comment:
Why are we using randomness and multiple threads here?  Why can't we just register a few hosts to the MLContext, and then call `StopExecution()` and ensure all the hosts have `IsCanceled` set to true?